File: src/main/java/me/syari/ss/discord/internal/entities/Member.java
package me.syari.ss.discord.internal.entities;

import me.syari.ss.discord.api.ISnowflake;
import me.syari.ss.discord.internal.JDA;
import me.syari.ss.discord.internal.utils.cache.SnowflakeReference;
import org.jetbrains.annotations.NotNull;

import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class Member implements ISnowflake {
    private final SnowflakeReference<Guild> guild;
    private final User user;
    private final Set<Role> roles = ConcurrentHashMap.newKeySet();

    private String nickname;
    private long boostDate;

    public Member(Guild guild, @NotNull User user) {
        JDA api = user.getJDA();
        this.guild = new SnowflakeReference<>(guild, api::getGuildById);
        this.user = user;
    }

    @NotNull
    public User getUser() {
        return user;
    }

    @NotNull
    private Guild getGuild() {
        return guild.resolve();
    }

    public String getNickname() {
        return nickname;
    }

    @NotNull
    public String getDisplayName() {
        return nickname != null ? nickname : getUser().getName();
    }

    @Override
    public long getIdLong() {
        return user.getIdLong();
    }

    public void setNickname(String nickname) {
        this.nickname = nickname;
    }

    public void setBoostDate(long boostDate) {
        this.boostDate = boostDate;
    }

    public Set<Role> getRoleSet() {
        return roles;
    }

    public long getBoostDateRaw() {
        return boostDate;
    }

    @Override
    public boolean equals(Object object) {
        if (object == this) {
            return true;
        }
        if (!(object instanceof Member)) {
            return false;
        }

        Member member = (Member) object;
        return member.user.getIdLong() == user.getIdLong() && member.guild.getIdLong() == guild.getIdLong();
    }

    @Override
    public int hashCode() {
        return (guild.getIdLong() + user.getId()).hashCode();
    }

    @Override
    public String toString() {
        return "MB:" + getDisplayName() + '(' + getUser().toString() + " / " + getGuild().toString() + ')';
    }

}

File: src/main/java/me/syari/ss/discord/internal/entities/Member.kt
package me.syari.ss.discord.internal.entities

import me.syari.ss.discord.api.ISnowflake
import me.syari.ss.discord.internal.utils.cache.SnowflakeReference
import java.util.function.LongFunction

class Member(guild: Guild?, user: User): ISnowflake {
    private val guild: SnowflakeReference<Guild?>
    val user: User
    var nickname: String? = null

    private fun getGuild(): Guild {
        return guild.resolve()
    }

    val displayName: String
        get() = (if (nickname != null) nickname else user.name) ?: "null"

    override val idLong: Long
        get() = user.idLong

    override fun equals(other: Any?): Boolean {
        if (other === this) return true
        if (other !is Member) return false
        return other.user.idLong == user.idLong && other.guild.idLong == guild.idLong
    }

    override fun hashCode(): Int {
        return (guild.idLong.toString() + user.idLong.toString()).hashCode()
    }

    override fun toString(): String {
        return "Member:" + displayName + '(' + user.toString() + " / " + getGuild().toString() + ')'
    }

    init {
        this.guild = SnowflakeReference(guild, LongFunction { id: Long -> user.api.getGuildById(id) })
        this.user = user
    }
}